Feat/task 03

numatay opened the pull request:
Closes Task-03

Main tasks:
1 - product-service serverless.yml contains configuration for 2 lambda functions, API is not working at all, but YAML configuration is correct - DONE

2 - The getProductsList OR getProductsById lambda function returns a correct response (POINT1) - **DONE**
The URL for getProductsList is https://p6cq8jgiid.execute-api.eu-west-1.amazonaws.com/products

3 - The getProductsById AND getProductsList lambda functions return a correct response code (POINT2) - **DONE**
The URL for getProductById is https://pnes1azyil.execute-api.eu-west-1.amazonaws.com/dev/products/5b718fac-8a4c-48b5-9379-215f1526598c

4 - Your own Frontend application is integrated with product service (/products API) and products from product-service are represented on Frontend. AND POINT1 and POINT2 are done. - **DONE**
The link to PR on frontend repo (rolling-scopes-school/nodejs-aws-fe#134)
The deployed version of the web app with backend integration is here: https://d5bzm0spt9d1j.cloudfront.net


Additional (optional) tasks (but nice to have):
+1 - Async/await is used in lambda functions - **DONE**
+1 - ES6 modules are used for product-service implementation - **DONE**
+1 - Webpack is configured for product-service - **DONE**
+1 - SWAGGER documentation is created for product-service - **NOT_DONE**
+1 - Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered) (You may use JEST) - **NOT_DONE**
+1 - Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase. - **DONE**
+1 - Main error scenarios are handled by API ("Product not found" error, try-catch blocks are used in lambda handlers). - **DONE**

Commit:
feat: Add 2 methods for products service

Commit:
feat: Add 2 methods for products service

Commit:
feat: Add complete implementation of getProductById and getAllProducts services

Commit:
feat: Add CORS headers and update seed

LoraBard commented on product-service/package.json:
```diff
@@ -0,0 +1,30 @@
+  "description": "Serverless webpack example using Typescript",
+  "main": "handler.js",
+  "scripts": {
+    "deploy": "serverless deploy -s dev -r eu-west-1 -v",
```
You can move region into serverless.ts  file (provider)

numatay replied:
Moved to `serverless.ts` and cleaned up `npm run deploy` script

LoraBard commented on product-service/serverless.ts:
```diff
@@ -0,0 +1,57 @@
+          http: {
+            method: 'get',
+            path: 'products/{productId}',
+            cors: true,
```
or  
provider 
  httpApi:
    cors: true
And use httpApi instead of http

numatay replied:
Moved `cors: true` to `httpApi` and reused `httpApi` with function

LoraBard commented on product-service/src/db.ts:
```diff
@@ -0,0 +1,38 @@
+
+    constructor() {
+        this.products = new Map<string, Product>();
+        Array.from(productsList.keys()).forEach(productKey => {
```
Possible. It looks a bit complicated. Did you think about complixety of this solution? Just:
private products = productsList;
getProducts(): Product[] {
return this.products;
}
getProductsById(id: string):Product {
  const product = this.products.find((item:Product) => item.id === id);
  ...
}

numatay replied:
Although this code looks confusing, I want to use a `map`  for the underlying data structures.
I would like to keep it this way if it doesn't affect the grading. ;)

LoraBard replied:
@numatay Of course, it doesn't affect the grading

LoraBard commented on product-service/src/db.ts:
```diff
@@ -0,0 +1,38 @@
+        });
+    }
+
+    async getProducts() {
```
Why async (without await)? Just use Promise.resolve() or create response handler function

numatay replied:
Removed `async` keyword from DB service and changed return value with `Promise.resolved` wrapped.

LoraBard commented on product-service/src/getAllProducts.ts:
```diff
@@ -0,0 +1,14 @@
+export const getAllProducts = async (_event, _context) => {
+    const products = await db.getProducts();
+    return {
+        statusCode: 200,
+        headers: {
+            'Access-Control-Allow-Origin': '*',
+            'Access-Control-Allow-Credentials': true,
+        },
+        body: JSON.stringify(products)
```
Duplicate of code. Please move to separate function

numatay replied:
Extracted to `addCorsHeaders` utilities and reused it in both functions.

LoraBard commented on product-service/src/getProductById.ts:
```diff
@@ -0,0 +1,26 @@
+export const getProductById = async (event, _context) => {
+
+    const { pathParameters } = event;
+    console.log("Event:", event, pathParameters);
```
remove console.log

numatay replied:
Cleaned up code.

LoraBard commented on product-service/src/getAllProducts.ts:
```diff
@@ -0,0 +1,14 @@
+import db from './db';
+import 'source-map-support/register';
+
+export const getAllProducts = async (_event, _context) => {
```
Please add types

numatay replied:
Added return types for the functions (`APIGatewayProxyHandler` from `aws-lambda` library)

LoraBard commented on product-service/src/getProductById.ts:
```diff
@@ -0,0 +1,26 @@
+import db from "./db";
+
+export const getProductById = async (event, _context) => {
```
the same

numatay replied:
Added return types for the functions (`APIGatewayProxyHandler` from `aws-lambda` library)

Commit:
refactor: the codebase accoring to issues in review

Commit:
refactor: the codebase accoring to issues in review

alexeyMohnatkin commented:
getProductById responds with 404 for any id

numatay commented:
> getProductById responds with 404 for any id

I must have messed up when deployed and updated the stack. Will fix it in a moment!

@alexeyMohnatkin, I updated the URL for `getProductById` in the description. Please have a look at it.

Commit:
fix: change to http for getProductById function
